### ISSUE TITLE
Behaviour of beforeEach and afterEach

### ISSUE THREAD

**Issue by jkimbo** [sindresorhus/ava#295]
I came across some weird behaviour today with ava where the following test gives the following output:

```js
import test from 'ava';

test.beforeEach(() => {
  console.log('beforeEach');
});

test.afterEach(() => {
  console.log('afterEach');
});

test('Test 1', (t) => {
  t.pass();
});

test('Test 2', (t) => {
  t.pass();
});
```
```bash
beforeEach
beforeEach
afterEach
afterEach
  ✔ Test 1
  ✔ Test 2

  2 tests passed
```

It was my assumption that `afterEach` would be called after a test and **before** the `beforeEach` of the next test. Is this expected behaviour? (I am using 0.7 btw)

Thanks!

**Comment by vdemedes** [comment]
Yes, `beforeEach` and `afterEach` work just like you expected, no worries ;) You are seeing such output, because AVA runs tests in parallel. It means, that `Test 2` does not wait for `Test 1` to complete, but both tests run at the same time.

**Comment by jkimbo** [comment]
@vdemedes ah ok that makes sense! I missed the difference between the test files being run in parallel and the tests being run concurrently. Since the tests get run concurrently how would you recommend patching global objects for tests? The original reason I came across this issue was because I was trying to patch the `process.nextTick` method so that I could make Promises synchronous in my tests however the `beforeEach` (where I patched the method) was being run before the corresponding `afterEach` had a chance to tear down the patch.

**Comment by vdemedes** [comment]
AVA both runs test files and tests concurrently :)

> Since the tests get run concurrently how would you recommend patching global objects for tests?

I'd recommend use `before()` and `after()` methods, here's the [docs for them](https://github.com/sindresorhus/ava#before--after-hooks).

> so that I could make Promises synchronous in my tests

I may not understand your goal correctly, but if I advise using [async/await](https://github.com/sindresorhus/ava#async-function-support) in your tests in order to write "synchronous looking" code.

Let me know if you have any more questions ;)

**Comment by jamestalmage** [comment]
@jkimbo 
You have two options.

Place the `process.nextTick` patch in `before` and `after`. These only get run once - so they shouldn't clobber each other. 

If your mock needs to maintain context, use [`t.context`](https://github.com/sindresorhus/ava#before--after-hooks) to pass information between `beforeEach` and the actual test (don't store it as a variable outside the functions). Note that `t.context` is not available in `before`/`after`, only the `Each` variants.

Use `t.serial` for your tests that require these mocks. If you have any async IO in your tests (i.e. disk/network requests), then you will end up sacrificing some speed. Otherwise, it should be easy.

**Comment by jamestalmage** [comment]
@vdemedes beat me by six seconds!

With a much better point regarding using `async`/`await` to boot.

**Comment by vdemedes** [comment]
![](http://i.giphy.com/10cRi957Dc7tM4.gif)

**Comment by jkimbo** [comment]
Haha thanks guys! Very helpful!

To give some more context around why I'm patching `process.nextTick`, I'm running some code that triggers a fetch request. I patch fetch out to be a promise I control in the test and then I assert some state, resolve the promise and assert the resultant state. The problem I had with that workflow was that the promise resolve would queue up a microtask and so I couldn't assert the next state synchronously. For the readability of the test I patch the `nextTick` method, store up the tasks queued up and run them at once so that I don't have to make the test async. Some sudo code that hopefully makes it clearer (React test):

```js
// mock fetch
global.fetch = sinon.stub();
const deferred = defer(); // create a deferred promise
fetch.returns(deferred.promise);

TestUtils.Simulate.click(ReactDOM.findDOMNode(stub.refs.button));

assert(fetch.calledOnce);
assert.equal(stub.state.searching, true);

// return a response
deferred.resolve({
  'status': 200,
  'json': () => ({
    'foo': 'bar',
  }),
});

runAllTicks(); // this runs all the tasks queued 

assert.equal(stub.state.searching, false);
```

Anyway I hope that explains why I can't use async/await etc.

Thanks again for the help!

**Comment by jkimbo** [comment]
FYI I got this idea from jest: https://facebook.github.io/jest/docs/api.html#jest-runallticks

**Comment by jamestalmage** [comment]
@jkimbo - I don't know a lot about React, so I may be missing something.

Why can't you just do:

```js
deferred.resolve({...});
 
await deferred.promise;

assert.equal(stub.state.searching, false);
```

If that returns too quickly (i.e. await returns before state has settled), you could do:

```js
import delay from 'delay';
// ...
await deferred.promise.then(delay(10)); // or delay(0), or delay(100), etc.

**Comment by jkimbo** [comment]
@jamestalmage that works really well thanks!
